src/NzbDrone.Core/MediaFiles/EpisodeImport/Aggregation/Aggregators/Augmenters/Quality/AugmentQualityFromReleaseName.cs

```diff
@@ -0,0 +1,89 @@
+using NzbDrone.Core.Parser.Model;
+using NzbDrone.Core.Qualities;
+
+namespace NzbDrone.Core.MediaFiles.EpisodeImport.Aggregation.Aggregators.Augmenters.Quality
+{
+    public class AugmentQualityFromReleaseName : IAugmentQuality
+    {
+        private readonly IDownloadHistoryService _downloadHistoryService;
+
+        public AugmentQualityFromReleaseName(IDownloadHistoryService downloadHistoryService)
+        {
+            _downloadHistoryService = downloadHistoryService;
+        }
+
+        public AugmentQualityResult AugmentQuality(LocalEpisode localEpisode, DownloadClientItem downloadClientItem)
+        {
+            // Don't try to augment if we can't lookup the grabbed history by downloadId
+            if (downloadClientItem == null)
+            {
+                return null;
+            }
+
+            var fileQuality = localEpisode.FileEpisodeInfo?.Quality;
+            var folderQuality = localEpisode.FolderEpisodeInfo?.Quality;
+            var localQuality = folderQuality ?? fileQuality;
+
+            if (localQuality == null)
+            {
+                return null;
+            }
+
+            // Return early if the file or folder quality source was parsed from the name.
+            if (localQuality.SourceDetectionSource == QualityDetectionSource.Name &&
+                localQuality.ResolutionDetectionSource == QualityDetectionSource.Name)
+            {
+                return null;
+            }
+
+            var history = _downloadHistoryService.GetLatestGrab(downloadClientItem.DownloadId);
+
+            if (history == null)
+            {
+                return null;
+            }
+
+            var historyQuality = QualityParser.ParseQuality(history.SourceTitle);
+
+            // Only return a source and/or resolution if the release name parsing matched via the name and the
+            // local folder/file did not. This way we'll override filenames that don't have the proper source/resolution
+            // with the source/resolution from the release name, but not attempt to override those that do.
+
+            var source = localQuality.SourceDetectionSource != QualityDetectionSource.Name &&
+                         historyQuality.SourceDetectionSource == QualityDetectionSource.Name
+                ? historyQuality.Quality.Source
+                : QualitySource.Unknown;
+
+            var resolution = localQuality.ResolutionDetectionSource != QualityDetectionSource.Name &&
+                              historyQuality.ResolutionDetectionSource == QualityDetectionSource.Name
+                ? historyQuality.Quality.Resolution
+                : (int?)null;
+
+            var revision = historyQuality.SourceDetectionSource == QualityDetectionSource.Name &&
+                           historyQuality.ResolutionDetectionSource == QualityDetectionSource.Name
+                ? historyQuality.Revision
+                : null;
+
+            if (source != QualitySource.Unknown && resolution.HasValue)
```

Comment:
Instead of deciding what we should return should we add a new confidence level and let `AggregateQuality` make the correct decision? That way we're not over complicating things here, we're just returning what we know from the release name.

Reply:
Adding a confidence level is not ideal because then you'd be combining confidence with priority (something we're already doing in AugmentQualityFromReleaseName to a limited degree).

Perhaps we should have AggregateQuality handle the 'equal confidence' scenario by preferring Folder over File over Release. Meaning that AugmentQualityFromReleaseName won't have to sort that out.

Technically, a conflicting Folder & File at the same confidence should produce an error. We don't now, and i'm ok with that. But it also means we're not going for the ideal solution anyway.

In practice, FolderInfo already contains the release title as provided by the indexer, so AugmentQualityFromReleaseName is only there to cover the scenario where it isn't available. So perhaps we should simplify it to deal solely with that scenario.
The separation of Source & Resolution already deals with a few of the other edge cases that we were trying to solve.

So I'm seeing the following options:
- either keep AugmentQualityFromReleaseName simple, based only on releasename and ignoring folder&file entirely
- or; adjust AugmentQualityFromReleaseName to return a result only if FolderInfo is absent
- or; adjust AugmentQualityFromReleaseName to return a result if Folder/File quality doesn't match the Grab quality as proposed earlier

Perhaps it's easier to simply sort the Augment* instances by a constant priority rather than SourceConfidence. If we always run through them by Folder, File, MediaInfo?, ReleaseName, Extension then the existing comparison based on confidence then ensures the right source/resolution is selected. Regardless of the chosen option above.

Reply:
> In practice, FolderInfo already contains the release title as provided by the indexer

For usenet yes, but not not always for torrents. We sometimes see torrent indexers change the name to include the source in the release’s name, even BTN does this for some web/webrip releases in this case the file and folder name different.

1. This could override the webrip/web DL naming from the file if the release name altered that, but feels like the best way to do it, we just return what we know and let something else figure out what to do with it.

2. Not feasible given release name doesn’t always equal folder name.

3. Similar to 1, we could end up overriding what the file/folder explicitly says, but I feel like taking in other values and using them to drive what return muddies the water.

A specific order would help so we don’t override something we prefer more, coupled with (1) I think it solves the problems with overriding something explicit.